feat: add operations field to mutation ApplyTo spec

pikilisaikiran opened the pull request:
# Add operations field to mutation ApplyTo specification

## Summary

This PR introduces an `operations` field to the `ApplyTo` specification in Gatekeeper mutations, enabling granular control over which Kubernetes admission operations (CREATE, UPDATE, DELETE) trigger mutations.

## Problem Solved

Fixes #4121: CronJob Pods getting stuck due to mutations attempting to modify immutable fields during UPDATE operations.

**Before:** Mutations always applied on both CREATE and UPDATE operations, causing failures when trying to mutate immutable Pod fields like environment variables.

**After:** Users can specify `operations: ["CREATE"]` to only mutate during resource creation, avoiding immutable field issues.

## Key Changes

### Core Implementation
- ✅ Add `Operations []string` field to `ApplyTo` struct
- ✅ Update mutation matching logic in `MatchWithApplyTo` function
- ✅ Add `Operation` field to `Mutable` type for context propagation
- ✅ Maintain 100% backward compatibility (defaults to CREATE+UPDATE)

### Testing & Validation
- ✅ Comprehensive unit tests for operation matching logic
- ✅ Live cluster testing with existing CRDs (no breaking changes)
- ✅ Multiple test scenarios covering all operation combinations

### Documentation & Examples
- ✅ Updated main mutation documentation with operations field explanation
- ✅ Added practical examples showing real-world usage patterns
- ✅ Created comprehensive README for mutation examples
- ✅ Enhanced Go code comments for better developer experience

## Usage Examples

### Solving Immutable Field Issues (Main Use Case)
```yaml
apiVersion: mutations.gatekeeper.sh/v1
kind: Assign
metadata:
  name: fix-cronjob-env-vars
spec:
  applyTo:
  - groups: [""]
    kinds: ["Pod"]
    versions: ["v1"]
    operations: ["CREATE"]  # Only apply on creation
  location: "spec.containers[name:*].env[name:MY_VAR].value"
  parameters:
    assign:
      value: "production"
```

### Other Supported Patterns
- `operations: ["CREATE"]` - Creation-only mutations
- `operations: ["UPDATE"]` - Update-only mutations  
- `operations: ["CREATE", "UPDATE"]` - Explicit default behavior
- `operations: ["DELETE"]` - Advanced cleanup scenarios
- No operations field - Backward compatible (CREATE+UPDATE)

## Backward Compatibility

✅ **Zero Breaking Changes**
- Existing mutations without `operations` field work unchanged
- Default behavior preserved as CREATE+UPDATE operations
- No required updates to existing deployments
- Gradual adoption - users can migrate at their own pace

## Testing Performed

1. **Unit Tests**: All new operation matching logic
2. **Integration Tests**: Live cluster deployment with existing CRDs
3. **Backward Compatibility**: Verified existing mutations unchanged
4. **Schema Validation**: Confirmed optional field works correctly
5. **Multiple Mutation Types**: Tested Assign, ModifySet, etc.

## Related Issue

Closes #4121

## Checklist

- [x] Added unit tests for new functionality
- [x] Updated documentation
- [x] Verified backward compatibility
- [x] Added practical examples
- [x] Tested on live cluster
- [x] No breaking changes introduced

Commit:
feat: add operations field to mutation ApplyTo spec

Add granular operation control to Gatekeeper mutations by introducing
an 'operations' field in the ApplyTo specification. This allows users
to specify which Kubernetes admission operations (CREATE, UPDATE, DELETE)
should trigger mutations.

Key Changes:
- Add Operations []string field to ApplyTo struct
- Update mutation matching logic to check operation compatibility
- Add Operation field to Mutable type for context propagation
- Maintain backward compatibility (defaults to CREATE+UPDATE)
- Add comprehensive unit tests for operation matching
- Update documentation with usage examples and best practices
- Include practical examples demonstrating various operation patterns

Fixes: Resolves issues with mutations attempting to modify immutable
fields during UPDATE operations, such as Pod environment variables
in CronJob scenarios.

Backward Compatibility: Existing mutations without operations field
continue working unchanged with default CREATE+UPDATE behavior.

Testing: Added unit tests, live cluster validation, and comprehensive
examples covering all supported operation combinations.

Signed-off-by: sai kiran pikili <saikiranpikili@localhost.localdomain>

JaydipGabani commented:
I am planning to discuss this in today's community meeting. I will take a look today to see if anything is missing. Thanks for the ping and thr PR.

chewong reviewed (commented):
Overall looks good!

JaydipGabani committed:
fixing docs, addressing feedback

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

JaydipGabani committed:
fix: address lint errors for ApplyTo receiver and imports

- Change ApplyTo method receivers from value to pointer to fix gocritic
  hugeParam warning (96 bytes struct)
- Fix import ordering (gci) in gator/reader and gator/verify packages
- Fix trailing whitespace in mutable.go

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

JaydipGabani committed:
updating API to only add operations in mutations, as well as reuse admission operations from k8s

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

JaydipGabani committed:
keeping behavior same with only supporting CREATE, UPDATE operations in mutation

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

aramase commented on pkg/mutation/match/apply_to.go:
```diff
@@ -55,3 +90,47 @@ func (a ApplyTo) Matches(gvk schema.GroupVersionKind) bool {
+		}
+	}
+	return false
+}
```
```suggestion
	// Check if the operation is explicitly allowed by the user
	return slices.Contains(a.Operations, operation)
}
```

this can be simplified

JaydipGabani replied:
updated to use slices.

aramase reviewed (changes requested):
First pass.

JaydipGabani committed:
fix: address review feedback for mutation operations field

- Use slices.Contains instead of manual loop
- Add test for empty operation string (audit/expansion context)
- Add OperationAll (*) wildcard support
- Add ValidateOperations for runtime validation

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

JaydipGabani commented on pkg/mutation/match/apply_to_test.go:
```diff
@@ -0,0 +1,265 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func TestApplyTo_MatchesOperation(t *testing.T) {
```
Added tests for this and wildcard as well.

JaydipGabani commented on pkg/mutation/match/apply_to.go:
```diff
@@ -55,3 +90,47 @@ func (a ApplyTo) Matches(gvk schema.GroupVersionKind) bool {
+// work in ApplyTo - empty means match all).
+// If operation is empty (e.g., in audit/expansion contexts), returns true
+// to maintain backward compatibility with non-admission flows.
+func (a *MutationApplyTo) MatchesOperation(operation admissionv1.Operation) bool {
```
Added wildcard support.

JaydipGabani committed:
Merge branch 'master' into feat/mutation-operations-field

aramase commented on pkg/mutation/match/apply_to.go:
```diff
@@ -55,3 +94,67 @@ func (a ApplyTo) Matches(gvk schema.GroupVersionKind) bool {
+// operations list. If no operations are specified, all operations are allowed
+// for backward compatibility (consistent with how empty groups/versions/kinds
+// work in ApplyTo - empty means match all).
+// If operation is empty (e.g., in audit/expansion contexts), returns true
```
The comment "consistent with how empty groups/versions/kinds work (empty = match all)" is inaccurate? I'm looking at

https://github.com/open-policy-agent/gatekeeper/blob/f30fd1adceff1c7b9d88867b28cd00b90999d1c8/pkg/mutation/mutators/core/mutator.go#L35-L39




```suggestion
// MatchesOperation returns true if the operation matches. If no operations 
// are specified, all operations are allowed for backward compatibility.
```

aramase commented on pkg/mutation/match/apply_to.go:
```diff
@@ -55,3 +94,67 @@ func (a ApplyTo) Matches(gvk schema.GroupVersionKind) bool {
+	for i, apply := range applyTo {
+		for _, op := range apply.Operations {
+			if !validOperations[op] {
+				return fmt.Errorf("invalid operation %q in applyTo[%d].operations: must be one of CREATE, UPDATE, DELETE, CONNECT, *", op, i)
```
Instead of returning early, collate all the errors and send back the string of error list. That's what we do in Kubernetes validations. Here is an example: https://github.com/kubernetes/kubernetes/blob/2fb84cfed3e0da83eec85dcc97d14db9b06915c9/staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go#L158-L184

aramase commented on pkg/mutation/types/mutable.go:
```diff
@@ -19,4 +20,7 @@ type Mutable struct {
 	// Source specifies which types of resources the mutator should be applied to
 	Source SourceType
+
+	// Operation is the admission operation being performed (CREATE, UPDATE, DELETE)
```
```suggestion
	// Operation is the admission operation being performed (CREATE, UPDATE, DELETE, CONNECT)
```

aramase reviewed (changes requested):
Another pass.

JaydipGabani committed:
fix: address review feedback on mutation operations

- Combine GVK and operation matching per applyTo entry to prevent
  cross-entry false positives
- Collate validation errors instead of returning early
- Reject wildcard * combined with other operations
- Use sets.Set[T] for validOperations
- Fix inaccurate comments; add CONNECT to Operation doc

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

JaydipGabani committed:
removing dead code

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

JaydipGabani committed:
test: use strings.Contains instead of custom helpers

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

sozercan committed:
Merge branch 'master' into feat/mutation-operations-field

JaydipGabani committed:
adding more tests

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

JaydipGabani committed:
add test for AppliesGVKAndOperation cross-entry matching

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

aramase reviewed (changes requested):
Another pass.

JaydipGabani committed:
fix: remove dead code, add duplicate op validation, document DELETE/CONNECT limitations

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

sozercan commented on pkg/mutation/mutators/core/mutator.go:
```diff
@@ -31,13 +31,18 @@ func NewTester(name string, kind string, path parser.Path, ptests []unversioned.
 // error if the applies are invalid.
-func NewValidatedBindings(name string, kind string, applies []match.ApplyTo) ([]schema.GroupVersionKind, error) {
+func NewValidatedBindings(name string, kind string, applies []match.MutationApplyTo) ([]schema.GroupVersionKind, error) {
 	for _, applyTo := range applies {
```
This now validates `operations`, but the schema-conflict path still appears to drop that scope: `NewValidatedBindings`/`SchemaBindings()` still expose only `[]GroupVersionKind`, and `pkg/mutation/schema.DB` keys implicit schemas by GVK. That means two mutators for the same GVK/path but disjoint operations (for example CREATE-only vs UPDATE-only) will still be reported as schema conflicts even though they can never both apply to the same admission request.

Can we plumb operation-aware schema bindings through here as well, e.g. key/group the schema DB by `(GVK, operation)` with empty/`*` expanded to the effective operation set, and add a regression test for disjoint operations not conflicting?

JaydipGabani replied:
fixed it! added a operation aware conflict detection.

JaydipGabani committed:
Fix operation-scoped mutation schema conflicts

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

JaydipGabani committed:
Propagate operations through expansion mutation, fix ci and generate manifests

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

JaydipGabani committed:
Merge branch 'master' into feat/mutation-operations-field

JaydipGabani committed:
Avoid allocation in operation matching

Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

sozercan commented on pkg/mutation/match/apply_to.go:
```diff
@@ -55,3 +92,79 @@ func (a ApplyTo) Matches(gvk schema.GroupVersionKind) bool {
+			admissionv1.Create,
+			admissionv1.Update,
+			admissionv1.Delete,
+			admissionv1.Connect,
```
**`EffectiveOperations()` registers DELETE/CONNECT schema bindings that can disable a working CREATE/UPDATE mutator**

This returns `{Create, Update, Delete, Connect}` for the empty/`*` case (and the raw list otherwise). The schema-conflict DB is keyed by `{GVK, Operation}`, so a mutator scoped to `operations: ["DELETE"]` — an op the mutation webhook never executes — still creates a `{GVK, Delete}` binding. If it shares a GVK+path with a default mutator (which also gets a `Delete` binding here) and their implied schemas disagree, `HasConflicts` disables **both** — including the normal CREATE/UPDATE mutator that does real work.

Validated on a kind cluster: an active default `Assign` + a `DELETE`-only `Assign` with conflicting paths → the active one reports `enforced 0/4` with `ErrConflictingSchema`, and a live Pod CREATE comes back unmutated.

Suggested fix: clamp effective operations to the set the webhook actually runs (`{CREATE, UPDATE}`) so DELETE/CONNECT-only mutators produce zero bindings and can't participate in conflict detection.

sozercan replied:
Addressed in a7b02ecd7. `EffectiveOperations()` (and the new `supportedMutationOperations` set) now clamp bindings to `{CREATE, UPDATE}`, so a DELETE/CONNECT-only mutator produces zero schema bindings and can no longer collide with an active CREATE/UPDATE mutator. Verified on a kind cluster: the active default mutator that previously went `enforced 0/4` with `ErrConflictingSchema` now reports `enforced 4/4` and mutates on a live CREATE.

sozercan commented on pkg/expansion/system.go:
```diff
@@ -202,6 +203,7 @@ func (s *System) expand(base *mutationtypes.Mutable) ([]*Resultant, error) {
 			Namespace: base.Namespace,
 			Username:  base.Username,
 			Source:    mutationtypes.SourceTypeGenerated,
+			Operation: base.Operation,
```
**expansion mutation reuses the parent admission operation**

The resultant `Mutable` passed to `mutationSystem.Mutate` inherits `base.Operation` (the parent generator's op, set from `req.Operation` in `policy.go`). A generated/expanded resource isn't actually undergoing that admission operation, so an operation-scoped mutator (e.g. `["CREATE"]`) fires on the preview based on the parent's op instead of being treated as operation-less.

Validated: the expansion unit test asserting a CREATE-only mutator is *not* applied during expansion fails against this code (the resultant gets mutated).

Suggested fix: don't propagate `Operation` into the expansion-path `Mutable`s (audit and gator already expand with no operation). The resultant *review* keeps `req.Operation` via `createReviewForResultant`, which is correct and separate.

sozercan replied:
Addressed in a7b02ecd7. `Operation` is no longer propagated into the expansion-path `Mutable`s in `policy.go`/`expansion/system.go`, so generated resources are mutated with no operation (matching audit and gator). The resultant *review* still keeps `req.Operation` via `createReviewForResultant`, which is correct. The expansion unit test asserting a CREATE-only mutator is not applied during expansion now passes.

sozercan commented on pkg/mutation/match/apply_to.go:
```diff
@@ -55,3 +92,79 @@ func (a ApplyTo) Matches(gvk schema.GroupVersionKind) bool {
+// If operation is empty (e.g., in audit/expansion contexts), only mutators
+// without operation filtering are allowed for backward compatibility.
+func (a *MutationApplyTo) MatchesOperation(operation admissionv1.Operation) bool {
+	if operation == "" {
```
**`["*"]`, `["CREATE","UPDATE"]`, and omitted are not interchangeable in `operation==""` paths**

In this `operation == ""` branch (audit, expansion, gator — no admission op), only an empty list or `*` returns true. An explicit `["CREATE","UPDATE"]` — which enumerates every op the webhook supports — returns false, so a user "being explicit" silently loses audit/expansion coverage that an otherwise-identical omitted-field mutator keeps.

Validated with `gator expand`: omitted and `["*"]` apply; `["CREATE","UPDATE"]` is dropped. On a live CREATE all three apply, so the divergence is confined to the operation-less paths.

Suggested fix: in the `operation==""` branch, treat a list covering all supported ops as equivalent to `*`/omitted. A genuinely partial list like `["CREATE"]` should still be excluded.

sozercan replied:
Addressed in a7b02ecd7. `MatchesOperation` now routes the `operation==""` case through `matchesAllSupportedMutationOperations()`, so a list covering all supported ops (`[CREATE,UPDATE]`) is treated the same as `*`/omitted. A genuinely partial scope like `[CREATE]` is still excluded. Verified with `gator expand`: omitted, `*`, and `[CREATE,UPDATE]` now all apply; `[CREATE]`-only stays correctly excluded.

sozercan commented on website/docs/mutation.md:
```diff
@@ -70,6 +71,36 @@ Note that the `applyTo` field is required for all mutators except `AssignMetadat
+**Common Use Cases:**
+- `operations: ["CREATE"]` - Ideal for setting initial values that shouldn't change on updates (e.g., environment variables with immutable constraints)
+- `operations: ["UPDATE"]` - For mutations that should only apply when resources are modified
+- `operations: ["CREATE", "UPDATE"]` - Apply to both creation and updates (this is what the webhook currently supports)
```
**docs describe behavior that doesn't match the code**

Two claims in this section are inaccurate given the operation semantics:
- (above, ~L95) "If `operations` is not specified or is empty, no operation filtering is applied" — at the binding layer, empty maps to `{CREATE, UPDATE}`, not "no filtering."
- This line presents `["CREATE","UPDATE"]` as equivalent to the default, but per Issue 3 it is dropped in audit/expansion.

Suggested fix: state that omitted / `*` / `["CREATE","UPDATE"]` are equivalent and cover the webhook-supported ops (CREATE, UPDATE); mark DELETE/CONNECT as accepted-but-inert; and note that mutators scoped to only *some* supported ops do not apply in audit/expansion.

sozercan replied:
Addressed in a7b02ecd7. `mutation.md` was rewritten: it now states that omitted / `*` / `[CREATE,UPDATE]` are equivalent and cover the webhook-supported ops (CREATE, UPDATE), marks DELETE/CONNECT as accepted-but-reserved (inert), and clarifies that mutators scoped to only some supported ops do not apply in audit/expansion while those covering all supported ops do.

sozercan reviewed (commented):
see comments

sozercan committed:
fix: scope mutation operations to CREATE/UPDATE and fix expansion op reuse

Signed-off-by: Sertac Ozercan <sozercan@gmail.com>

sozercan committed:
fix: reject DELETE/CONNECT mutation operations until the webhook supports them

The mutation webhook only executes CREATE and UPDATE, so accepting DELETE and
CONNECT as operations values created mutators that are valid but can never run,
and committed the API to silently activating them on a future upgrade. Reject
concrete DELETE/CONNECT at the CRD enum and in ValidateOperations; keep CREATE,
UPDATE, omitted, and "*". "*"/omitted continue to mean "all currently-supported
operations" and broaden only via a deliberate, release-noted change to the
supported set. Docs clarify the accepted values and the "*" broadening contract.

Signed-off-by: Sertac Ozercan <sozercan@gmail.com>

sozercan commented on pkg/mutation/match/apply_to.go:
```diff
@@ -21,9 +53,21 @@ type ApplyTo struct {
+type MutationApplyTo struct {
+	ApplyTo `json:",inline"`
+	// Operations specifies which admission operations (CREATE, UPDATE, DELETE, CONNECT, *) should trigger
+	// this mutation. If empty, all operations are allowed for backward compatibility.
```
Revisited this and went with your first option — restricting the accepted values rather than only documenting it. As of 5b81a7d45, concrete `DELETE` and `CONNECT` are now rejected for mutation `operations`, at two layers:

- the CRD enum is narrowed to `CREATE;UPDATE;*`, so the API server rejects them before persistence (verified on a kind cluster: `Unsupported value: "DELETE": supported values: "CREATE", "UPDATE", "*"`);
- `match.ValidateOperations` drops `Delete`/`Connect` from its accepted set as a webhook-side backstop that also survives CRD/controller version skew.

This avoids the valid-but-never-fires mutators you flagged, and — since rejecting a concrete value now is the backward-compatible direction (values can be added later, not removed) — it also avoids committing the API to silently activating those ops on a future upgrade.

`*` and an omitted field are kept and continue to mean "all operations the mutation webhook currently supports" (today CREATE/UPDATE). If support for more operations is ever added, `*`/omitted broaden via a deliberate, release-noted change to the supported set; users who want a fixed scope across upgrades can list operations explicitly. The field doc comment and website docs were updated to spell this out.

JaydipGabani reviewed (approved):
lgtm

JaydipGabani committed:
Merge branch 'master' into feat/mutation-operations-field

sozercan reviewed (approved):
lgtm

JaydipGabani committed:
Merge branch 'master' into feat/mutation-operations-field
